flocker/dockerplugin/_script.py

```diff
@@ -0,0 +1,93 @@
+# Copyright ClusterHQ Inc. See LICENSE file for details.
+
+"""
+Command to start up the Docker plugin.
+"""
+
+from os import umask
+from stat import S_IRUSR, S_IWUSR, S_IXUSR
+
+from twisted.python.usage import Options
+from twisted.internet.endpoints import serverFromString
+from twisted.application.internet import StreamServerEndpointService
+from twisted.web.server import Site
+from twisted.python.filepath import FilePath
+from twisted.internet.address import UNIXAddress
+
+from ..common.script import (
+    flocker_standard_options, FlockerScriptRunner, main_for_service)
+from ._api import VolumePlugin
+from ..node.script import get_configuration
+from ..apiclient import FlockerClient
+from ..control.httpapi import REST_API_PORT
+
+PLUGIN_PATH = FilePath("/run/docker/plugins/flocker/flocker.sock")
+
+
+@flocker_standard_options
+class DockerPluginOptions(Options):
+    """
+    Command-line options for the Docker plugin.
+    """
+    optParameters = [
+        ["rest-api-port", "p", REST_API_PORT,
+         "Port to connect to for control service REST API."],
+        ["agent-config", "c", "/etc/flocker/agent.yml",
+         "The configuration file for the local agent."],
+    ]
+
+    def postOptions(self):
+        self['agent-config'] = FilePath(self['agent-config'])
+
+
+class DockerPluginScript(object):
+    """
+    Start the Docker plugin.
+    """
+    def _create_listening_directory(self, directory_path):
+        """
+        Create the parent directory for the Unix socket if it doesn't exist.
+
+        :param FilePath directory_path: The directory to create.
+        """
+        original_umask = umask(0)
+        try:
+            if not directory_path.exists():
+                directory_path.makedirs()
+            directory_path.chmod(S_IRUSR | S_IWUSR | S_IXUSR)
+        finally:
+            umask(original_umask)
+
+    def main(self, reactor, options):
+        # Many places in both twisted.web and Klein are unhappy with
+        # listening on Unix socket, fix that by pretending we have a port
+        # number:
```

Comment:
https://twistedmatrix.com/trac/ticket/5406 though I suspect there's more to be hit in Klein itself once that is fixed. Will add link post-review.